### PR TITLE
system/utils/stackmonitor : Move terminated task logging func to kernel

### PR DESCRIPTION
**Pull request by jeongchanKim**
Previously, kernel used app/system resource(stkmon_arr) for logging terminated task information.
Now kernel has resource for logging, and stackmonitor in app/system copies the data when they need.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [d8247fed4f06deed80dd5f10409e39ea34f71787] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d8247fed4f06deed80dd5f10409e39ea34f71787] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fc5d7e0c89adeb6317f312ce0a6d15f88b00f81e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fc5d7e0c89adeb6317f312ce0a6d15f88b00f81e] - Code Rule Check (C++) OK.

File: apps/system/utils/utils_stackmonitor.c [hunk 1]
```diff
@@ -282,35 +292,7 @@ static void stackmonitor_stop(void)
...
-#endif
-#if (CONFIG_TASK_NAME_SIZE > 0)
-	for (name_idx = 0; name_idx < CONFIG_TASK_NAME_SIZE + 1; name_idx++) {
-		stkmon_arr[stkmon_chk_idx % (CONFIG_MAX_TASKS * 2)].chk_name[name_idx] = tcb->name[name_idx];
-	}
-#endif
-	stkmon_chk_idx %= (CONFIG_MAX_TASKS * 2);
-	stkmon_chk_idx++;
-}
+
```
Comment (drashti304) [inline apps/system/utils/utils_stackmonitor.c]: I think, this extra line is not necessary

File: apps/system/utils/Kconfig [hunk 2]
```diff
@@ -193,7 +193,7 @@ config ENABLE_STACKMONITOR
 if ENABLE_STACKMONITOR
 config STACKMONITOR_STACKSIZE
 	int "Stack monitor daemon stack size"
-	default 2048
+	default 4096
```
Comment (sunghan-chang) [inline apps/system/utils/Kconfig]: Why is this necessary?

File: os/kernel/task/task_prctl.c [hunk 3]
```diff
@@ -166,6 +166,16 @@ int prctl(int option, ...)
 	goto errout;
 #endif
 
+	case PR_COPY_STKMONLOG:
+	{
+#if defined(CONFIG_ENABLE_STACKMONITOR) && defined(CONFIG_DEBUG)
+		struct stkmon_save_s *dest_arr = va_arg(ap, struct stkmon_save_s *);
```
Comment (sunghan-chang) [inline os/kernel/task/task_prctl.c]: how about buf instead of arr?

File: os/kernel/task/task_prctl.c [hunk 4]
```diff
@@ -166,6 +166,16 @@ int prctl(int option, ...)
...
+		stkmon_copy_log(dest_arr);
+#else
+		sdbg("Not supported StackMonitor Logging\n");
```
Comment (sunghan-chang) [inline os/kernel/task/task_prctl.c]: err = EINVAL;

File: os/kernel/task/task_prctl.c [hunk 5]
```diff
@@ -166,6 +166,16 @@ int prctl(int option, ...)
...
+#endif
+	}
```
Comment (sunghan-chang) [inline os/kernel/task/task_prctl.c]: You should add return

File: os/kernel/sched/sched_stkmon_logging.c [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+/****************************************************************************
+ * Name: stkmon_logging
+ *
+ * Description:
+ *   This function saves the terminated task/pthread's information for
+ *  stack monitor.
+ *
+ ****************************************************************************/
+void stkmon_logging(struct tcb_s *tcb)
```
Comment (sunghan-chang) [inline os/kernel/sched/sched_stkmon_logging.c]: How about `sched_save_stackinfo_terminated`?

File: os/kernel/sched/sched_stkmon_logging.c [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Global Variables
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Variables
+ ****************************************************************************/
+static struct stkmon_save_s stkmon_arr[CONFIG_MAX_TASKS * 2];
```
Comment (sunghan-chang) [inline os/kernel/sched/sched_stkmon_logging.c]: terminated_thread_stackinfo_buf?

File: apps/system/utils/Kconfig [hunk 8]
```diff
@@ -193,7 +193,7 @@ config ENABLE_STACKMONITOR
...
 	---help---
 		The stack size of the stack monitor daemon.  Default: 2048 bytes
```
Comment (intom) [inline apps/system/utils/Kconfig]: If you will change 2048 to 4096, how about changing to 4096?

**Comment by seinfra** [comment]
Target : [d09cc798b9626212c775bf8c87d3992ecc469d93] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d09cc798b9626212c775bf8c87d3992ecc469d93] - Code Rule Check OK.

File: os/kernel/task/task_prctl.c (hunk 4 above)
Comment (jeongchanKim) [reply os/kernel/task/task_prctl.c]: I think ENOSYS is better, so add `err = ENOSYS`.

File: os/kernel/task/task_prctl.c (hunk 5 above)
Comment (jeongchanKim) [reply os/kernel/task/task_prctl.c]: Done.

File: apps/system/utils/Kconfig (hunk 2 above)
Comment (jeongchanKim) [reply apps/system/utils/Kconfig]: Previous stackmonitor reads the terminated thread's stack data directly. But for supporting protected build, stackmonitor should read the data through system call and function calls, so its stack should be increased.

**Comment by seinfra** [comment]
Target : [5d30368485591841a70e315056b122123645dce9] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5d30368485591841a70e315056b122123645dce9] - Code Rule Check OK.
